potential crash in vm.go

In vm.go code like this is used:

```
vp := util.NewValueParser(value)

		switch f.Name() {
		case "admin_reserve_kbytes":
			vm.AdminReserveKbytes = *vp.PInt64()
```

if the methods of valueparser return a nil pionter, which they do if they can't parse the value, this will generate a crash:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x51d9a3]
```

Since this was added quite recently I'd simply change the interface to use pointers for parsed values, thus simply resulting in nil pointers in the return struct fields.

--- a/vm.go
+++ b/vm.go
@@ -31,49 +31,49 @@ import (
 // Each file contains one line with a single numerical value, except lowmem_reserve_ratio which holds an array
 // and numa_zonelist_order (deprecated) which is a string
 type VM struct {
-	AdminReserveKbytes        int64   // /proc/sys/vm/admin_reserve_kbytes
-	BlockDump                 int64   // /proc/sys/vm/block_dump
-	CompactUnevictableAllowed int64   // /proc/sys/vm/compact_unevictable_allowed
-	DirtyBackgroundBytes      int64   // /proc/sys/vm/dirty_background_bytes
-	DirtyBackgroundRatio      int64   // /proc/sys/vm/dirty_background_ratio
-	DirtyBytes                int64   // /proc/sys/vm/dirty_bytes
-	DirtyExpireCentisecs      int64   // /proc/sys/vm/dirty_expire_centisecs
-	DirtyRatio                int64   // /proc/sys/vm/dirty_ratio
-	DirtytimeExpireSeconds    int64   // /proc/sys/vm/dirtytime_expire_seconds
-	DirtyWritebackCentisecs   int64   // /proc/sys/vm/dirty_writeback_centisecs
-	DropCaches                int64   // /proc/sys/vm/drop_caches
-	ExtfragThreshold          int64   // /proc/sys/vm/extfrag_threshold
-	HugetlbShmGroup           int64   // /proc/sys/vm/hugetlb_shm_group
-	LaptopMode                int64   // /proc/sys/vm/laptop_mode
-	LegacyVaLayout            int64   // /proc/sys/vm/legacy_va_layout
-	LowmemReserveRatio        []int64 // /proc/sys/vm/lowmem_reserve_ratio
-	MaxMapCount               int64   // /proc/sys/vm/max_map_count
-	MemoryFailureEarlyKill    int64   // /proc/sys/vm/memory_failure_early_kill
-	MemoryFailureRecovery     int64   // /proc/sys/vm/memory_failure_recovery
-	MinFreeKbytes             int64   // /proc/sys/vm/min_free_kbytes
-	MinSlabRatio              int64   // /proc/sys/vm/min_slab_ratio
-	MinUnmappedRatio          int64   // /proc/sys/vm/min_unmapped_ratio
-	MmapMinAddr               int64   // /proc/sys/vm/mmap_min_addr
-	NrHugepages               int64   // /proc/sys/vm/nr_hugepages
-	NrHugepagesMempolicy      int64   // /proc/sys/vm/nr_hugepages_mempolicy
-	NrOvercommitHugepages     int64   // /proc/sys/vm/nr_overcommit_hugepages
-	NumaStat                  int64   // /proc/sys/vm/numa_stat
-	NumaZonelistOrder         string  // /proc/sys/vm/numa_zonelist_order
-	OomDumpTasks              int64   // /proc/sys/vm/oom_dump_tasks
-	OomKillAllocatingTask     int64   // /proc/sys/vm/oom_kill_allocating_task
-	OvercommitKbytes          int64   // /proc/sys/vm/overcommit_kbytes
-	OvercommitMemory          int64   // /proc/sys/vm/overcommit_memory
-	OvercommitRatio           int64   // /proc/sys/vm/overcommit_ratio
-	PageCluster               int64   // /proc/sys/vm/page-cluster
-	PanicOnOom                int64   // /proc/sys/vm/panic_on_oom
-	PercpuPagelistFraction    int64   // /proc/sys/vm/percpu_pagelist_fraction
-	StatInterval              int64   // /proc/sys/vm/stat_interval
-	Swappiness                int64   // /proc/sys/vm/swappiness
-	UserReserveKbytes         int64   // /proc/sys/vm/user_reserve_kbytes
-	VfsCachePressure          int64   // /proc/sys/vm/vfs_cache_pressure
-	WatermarkBoostFactor      int64   // /proc/sys/vm/watermark_boost_factor
-	WatermarkScaleFactor      int64   // /proc/sys/vm/watermark_scale_factor
-	ZoneReclaimMode           int64   // /proc/sys/vm/zone_reclaim_mode
+	AdminReserveKbytes        *int64   // /proc/sys/vm/admin_reserve_kbytes
+	BlockDump                 *int64   // /proc/sys/vm/block_dump
+	CompactUnevictableAllowed *int64   // /proc/sys/vm/compact_unevictable_allowed
+	DirtyBackgroundBytes      *int64   // /proc/sys/vm/dirty_background_bytes
+	DirtyBackgroundRatio      *int64   // /proc/sys/vm/dirty_background_ratio
+	DirtyBytes                *int64   // /proc/sys/vm/dirty_bytes
+	DirtyExpireCentisecs      *int64   // /proc/sys/vm/dirty_expire_centisecs
+	DirtyRatio                *int64   // /proc/sys/vm/dirty_ratio
+	DirtytimeExpireSeconds    *int64   // /proc/sys/vm/dirtytime_expire_seconds
+	DirtyWritebackCentisecs   *int64   // /proc/sys/vm/dirty_writeback_centisecs
+	DropCaches                *int64   // /proc/sys/vm/drop_caches
+	ExtfragThreshold          *int64   // /proc/sys/vm/extfrag_threshold
+	HugetlbShmGroup           *int64   // /proc/sys/vm/hugetlb_shm_group
+	LaptopMode                *int64   // /proc/sys/vm/laptop_mode
+	LegacyVaLayout            *int64   // /proc/sys/vm/legacy_va_layout
+	LowmemReserveRatio        []*int64 // /proc/sys/vm/lowmem_reserve_ratio
+	MaxMapCount               *int64   // /proc/sys/vm/max_map_count
+	MemoryFailureEarlyKill    *int64   // /proc/sys/vm/memory_failure_early_kill
+	MemoryFailureRecovery     *int64   // /proc/sys/vm/memory_failure_recovery
+	MinFreeKbytes             *int64   // /proc/sys/vm/min_free_kbytes
+	MinSlabRatio              *int64   // /proc/sys/vm/min_slab_ratio
+	MinUnmappedRatio          *int64   // /proc/sys/vm/min_unmapped_ratio
+	MmapMinAddr               *int64   // /proc/sys/vm/mmap_min_addr
+	NrHugepages               *int64   // /proc/sys/vm/nr_hugepages
+	NrHugepagesMempolicy      *int64   // /proc/sys/vm/nr_hugepages_mempolicy
+	NrOvercommitHugepages     *int64   // /proc/sys/vm/nr_overcommit_hugepages
+	NumaStat                  *int64   // /proc/sys/vm/numa_stat
+	NumaZonelistOrder         string   // /proc/sys/vm/numa_zonelist_order
+	OomDumpTasks              *int64   // /proc/sys/vm/oom_dump_tasks
+	OomKillAllocatingTask     *int64   // /proc/sys/vm/oom_kill_allocating_task
+	OvercommitKbytes          *int64   // /proc/sys/vm/overcommit_kbytes
+	OvercommitMemory          *int64   // /proc/sys/vm/overcommit_memory
+	OvercommitRatio           *int64   // /proc/sys/vm/overcommit_ratio
+	PageCluster               *int64   // /proc/sys/vm/page-cluster
+	PanicOnOom                *int64   // /proc/sys/vm/panic_on_oom
+	PercpuPagelistFraction    *int64   // /proc/sys/vm/percpu_pagelist_fraction
+	StatInterval              *int64   // /proc/sys/vm/stat_interval
+	Swappiness                *int64   // /proc/sys/vm/swappiness
+	UserReserveKbytes         *int64   // /proc/sys/vm/user_reserve_kbytes
+	VfsCachePressure          *int64   // /proc/sys/vm/vfs_cache_pressure
+	WatermarkBoostFactor      *int64   // /proc/sys/vm/watermark_boost_factor
+	WatermarkScaleFactor      *int64   // /proc/sys/vm/watermark_scale_factor
+	ZoneReclaimMode           *int64   // /proc/sys/vm/zone_reclaim_mode
 }
 
 // VM reads the VM statistics from the specified `proc` filesystem.
@@ -109,97 +109,97 @@ func (fs FS) VM() (*VM, error) {
 
 		switch f.Name() {
 		case "admin_reserve_kbytes":
-			vm.AdminReserveKbytes = *vp.PInt64()
+			vm.AdminReserveKbytes = vp.PInt64()
 		case "block_dump":
-			vm.BlockDump = *vp.PInt64()
+			vm.BlockDump = vp.PInt64()
 		case "compact_unevictable_allowed":
-			vm.CompactUnevictableAllowed = *vp.PInt64()
+			vm.CompactUnevictableAllowed = vp.PInt64()
 		case "dirty_background_bytes":
-			vm.DirtyBackgroundBytes = *vp.PInt64()
+			vm.DirtyBackgroundBytes = vp.PInt64()
 		case "dirty_background_ratio":
-			vm.DirtyBackgroundRatio = *vp.PInt64()
+			vm.DirtyBackgroundRatio = vp.PInt64()
 		case "dirty_bytes":
-			vm.DirtyBytes = *vp.PInt64()
+			vm.DirtyBytes = vp.PInt64()
 		case "dirty_expire_centisecs":
-			vm.DirtyExpireCentisecs = *vp.PInt64()
+			vm.DirtyExpireCentisecs = vp.PInt64()
 		case "dirty_ratio":
-			vm.DirtyRatio = *vp.PInt64()
+			vm.DirtyRatio = vp.PInt64()
 		case "dirtytime_expire_seconds":
-			vm.DirtytimeExpireSeconds = *vp.PInt64()
+			vm.DirtytimeExpireSeconds = vp.PInt64()
 		case "dirty_writeback_centisecs":
-			vm.DirtyWritebackCentisecs = *vp.PInt64()
+			vm.DirtyWritebackCentisecs = vp.PInt64()
 		case "drop_caches":
-			vm.DropCaches = *vp.PInt64()
+			vm.DropCaches = vp.PInt64()
 		case "extfrag_threshold":
-			vm.ExtfragThreshold = *vp.PInt64()
+			vm.ExtfragThreshold = vp.PInt64()
 		case "hugetlb_shm_group":
-			vm.HugetlbShmGroup = *vp.PInt64()
+			vm.HugetlbShmGroup = vp.PInt64()
 		case "laptop_mode":
-			vm.LaptopMode = *vp.PInt64()
+			vm.LaptopMode = vp.PInt64()
 		case "legacy_va_layout":
-			vm.LegacyVaLayout = *vp.PInt64()
+			vm.LegacyVaLayout = vp.PInt64()
 		case "lowmem_reserve_ratio":
 			stringSlice := strings.Fields(value)
-			int64Slice := make([]int64, 0, len(stringSlice))
+			pint64Slice := make([]*int64, 0, len(stringSlice))
 			for _, value := range stringSlice {
 				vp := util.NewValueParser(value)
-				int64Slice = append(int64Slice, *vp.PInt64())
+				pint64Slice = append(pint64Slice, vp.PInt64())
 			}
-			vm.LowmemReserveRatio = int64Slice
+			vm.LowmemReserveRatio = pint64Slice
 		case "max_map_count":
-			vm.MaxMapCount = *vp.PInt64()
+			vm.MaxMapCount = vp.PInt64()
 		case "memory_failure_early_kill":
-			vm.MemoryFailureEarlyKill = *vp.PInt64()
+			vm.MemoryFailureEarlyKill = vp.PInt64()
 		case "memory_failure_recovery":
-			vm.MemoryFailureRecovery = *vp.PInt64()
+			vm.MemoryFailureRecovery = vp.PInt64()
 		case "min_free_kbytes":
-			vm.MinFreeKbytes = *vp.PInt64()
+			vm.MinFreeKbytes = vp.PInt64()
 		case "min_slab_ratio":
-			vm.MinSlabRatio = *vp.PInt64()
+			vm.MinSlabRatio = vp.PInt64()
 		case "min_unmapped_ratio":
-			vm.MinUnmappedRatio = *vp.PInt64()
+			vm.MinUnmappedRatio = vp.PInt64()
 		case "mmap_min_addr":
-			vm.MmapMinAddr = *vp.PInt64()
+			vm.MmapMinAddr = vp.PInt64()
 		case "nr_hugepages":
-			vm.NrHugepages = *vp.PInt64()
+			vm.NrHugepages = vp.PInt64()
 		case "nr_hugepages_mempolicy":
-			vm.NrHugepagesMempolicy = *vp.PInt64()
+			vm.NrHugepagesMempolicy = vp.PInt64()
 		case "nr_overcommit_hugepages":
-			vm.NrOvercommitHugepages = *vp.PInt64()
+			vm.NrOvercommitHugepages = vp.PInt64()
 		case "numa_stat":
-			vm.NumaStat = *vp.PInt64()
+			vm.NumaStat = vp.PInt64()
 		case "numa_zonelist_order":
 			vm.NumaZonelistOrder = value
 		case "oom_dump_tasks":
-			vm.OomDumpTasks = *vp.PInt64()
+			vm.OomDumpTasks = vp.PInt64()
 		case "oom_kill_allocating_task":
-			vm.OomKillAllocatingTask = *vp.PInt64()
+			vm.OomKillAllocatingTask = vp.PInt64()
 		case "overcommit_kbytes":
-			vm.OvercommitKbytes = *vp.PInt64()
+			vm.OvercommitKbytes = vp.PInt64()
 		case "overcommit_memory":
-			vm.OvercommitMemory = *vp.PInt64()
+			vm.OvercommitMemory = vp.PInt64()
 		case "overcommit_ratio":
-			vm.OvercommitRatio = *vp.PInt64()
+			vm.OvercommitRatio = vp.PInt64()
 		case "page-cluster":
-			vm.PageCluster = *vp.PInt64()
+			vm.PageCluster = vp.PInt64()
 		case "panic_on_oom":
-			vm.PanicOnOom = *vp.PInt64()
+			vm.PanicOnOom = vp.PInt64()
 		case "percpu_pagelist_fraction":
-			vm.PercpuPagelistFraction = *vp.PInt64()
+			vm.PercpuPagelistFraction = vp.PInt64()
 		case "stat_interval":
-			vm.StatInterval = *vp.PInt64()
+			vm.StatInterval = vp.PInt64()
 		case "swappiness":
-			vm.Swappiness = *vp.PInt64()
+			vm.Swappiness = vp.PInt64()
 		case "user_reserve_kbytes":
-			vm.UserReserveKbytes = *vp.PInt64()
+			vm.UserReserveKbytes = vp.PInt64()
 		case "vfs_cache_pressure":
-			vm.VfsCachePressure = *vp.PInt64()
+			vm.VfsCachePressure = vp.PInt64()
 		case "watermark_boost_factor":
-			vm.WatermarkBoostFactor = *vp.PInt64()
+			vm.WatermarkBoostFactor = vp.PInt64()
 		case "watermark_scale_factor":
-			vm.WatermarkScaleFactor = *vp.PInt64()
+			vm.WatermarkScaleFactor = vp.PInt64()
 		case "zone_reclaim_mode":
-			vm.ZoneReclaimMode = *vp.PInt64()
+			vm.ZoneReclaimMode = vp.PInt64()
 		}
 		if err := vp.Err(); err != nil {
 			return nil, err

--- a/vm_test.go
+++ b/vm_test.go
@@ -14,51 +14,72 @@
 package procfs
 
 import (
-	"reflect"
 	"testing"
+
+	"github.com/google/go-cmp/cmp"
 )
+
+func newPInt64(i int64) *int64 {
+	return &i
+}
 
 func TestVM(t *testing.T) {
 	fs, err := NewFS(procTestFixtures)
 	if err != nil {
 		t.Fatal(err)
 	}
-	stats, err := fs.VM()
+	got, err := fs.VM()
 	if err != nil {
 		t.Fatal(err)
 	}
-
-	lowmemreserveratio := []int64{256, 256, 32, 0, 0}
-	vm := &VM{
-		AdminReserveKbytes:        8192,
-		CompactUnevictableAllowed: 1,
-		DirtyBackgroundRatio:      10,
-		DirtyExpireCentisecs:      3000,
-		DirtyRatio:                20,
-		DirtytimeExpireSeconds:    43200,
-		DirtyWritebackCentisecs:   500,
-		ExtfragThreshold:          500,
-		LaptopMode:                5,
+	zeroPointer := newPInt64(0)
+	lowmemreserveratio := []*int64{newPInt64(256), newPInt64(256), newPInt64(32), zeroPointer, zeroPointer}
+	want := &VM{
+		AdminReserveKbytes:        newPInt64(8192),
+		BlockDump:                 zeroPointer,
+		CompactUnevictableAllowed: newPInt64(1),
+		DirtyBackgroundBytes:      zeroPointer,
+		DirtyBackgroundRatio:      newPInt64(10),
+		DirtyBytes:                zeroPointer,
+		DirtyExpireCentisecs:      newPInt64(3000),
+		DirtyRatio:                newPInt64(20),
+		DirtytimeExpireSeconds:    newPInt64(43200),
+		DirtyWritebackCentisecs:   newPInt64(500),
+		DropCaches:                zeroPointer,
+		ExtfragThreshold:          newPInt64(500),
+		HugetlbShmGroup:           zeroPointer,
+		LaptopMode:                newPInt64(5),
+		LegacyVaLayout:            zeroPointer,
 		LowmemReserveRatio:        lowmemreserveratio,
-		MaxMapCount:               65530,
-		MemoryFailureRecovery:     1,
-		MinFreeKbytes:             67584,
-		MinSlabRatio:              5,
-		MinUnmappedRatio:          1,
-		MmapMinAddr:               65536,
-		NumaStat:                  1,
+		MaxMapCount:               newPInt64(65530),
+		MemoryFailureEarlyKill:    zeroPointer,
+		MemoryFailureRecovery:     newPInt64(1),
+		MinFreeKbytes:             newPInt64(67584),
+		MinSlabRatio:              newPInt64(5),
+		MinUnmappedRatio:          newPInt64(1),
+		MmapMinAddr:               newPInt64(65536),
+		NumaStat:                  newPInt64(1),
 		NumaZonelistOrder:         "Node",
-		OomDumpTasks:              1,
-		OvercommitRatio:           50,
-		PageCluster:               3,
-		StatInterval:              1,
-		Swappiness:                60,
-		UserReserveKbytes:         131072,
-		VfsCachePressure:          100,
-		WatermarkBoostFactor:      15000,
-		WatermarkScaleFactor:      10,
+		NrHugepages:               zeroPointer,
+		NrHugepagesMempolicy:      zeroPointer,
+		NrOvercommitHugepages:     zeroPointer,
+		OomDumpTasks:              newPInt64(1),
+		OomKillAllocatingTask:     zeroPointer,
+		OvercommitKbytes:          zeroPointer,
+		OvercommitMemory:          zeroPointer,
+		OvercommitRatio:           newPInt64(50),
+		PageCluster:               newPInt64(3),
+		PanicOnOom:                zeroPointer,
+		PercpuPagelistFraction:    zeroPointer,
+		StatInterval:              newPInt64(1),
+		Swappiness:                newPInt64(60),
+		UserReserveKbytes:         newPInt64(131072),
+		VfsCachePressure:          newPInt64(100),
+		WatermarkBoostFactor:      newPInt64(15000),
+		WatermarkScaleFactor:      newPInt64(10),
+		ZoneReclaimMode:           zeroPointer,
 	}
-	if !reflect.DeepEqual(vm, stats) {
-		t.Errorf("Result not correct: want %v\n, have %v", vm, stats)
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Fatalf("unexpected power supply class (-want +got):\n%s", diff)
 	}
 }

